### ISSUE TITLE
Reproduction Issue with Resnet and ViT8

### ISSUE THREAD

**Issue by FrsECM** [rbalestr-lab/lejepa#20]
Hello,
First of all thank you for your paper that is very well written and easy to understand. Concepts are simple, very well justified that was very pleasant to read.

I followed your tutorial and implemented it in order to reproduce your results on Imagenette (INET10).
Due to my work environment, i'd like to pretrain convolution network, that's why i did an implementation for a resnet before the one with vit8.

First i implemented models : 
<details>
<summary>Model Implementation</summary> 

```python
class ResNetLeJEPA(nn.Module):
    def __init__(self,backbone:str='resnet18',proj_dim:int=256):
        super(ResNetLeJEPA,self).__init__()
        self.backbone = backbones[backbone]()
        if hasattr(self.backbone,'fc'):
            # Remove the final classification layer of ResNet
            input_dim = self.backbone.fc.in_features
            self.backbone.fc = nn.Identity()
        
        self.proj = MLP(input_dim,[2048,2048,proj_dim],norm_layer=nn.BatchNorm1d)
    
    def forward(self,x:torch.Tensor)->torch.Tensor:
        if x.ndim == 5:
            # Perform inference on views
            b,v,c,h,w = x.size()
            x = x.view(b*v,c,h,w)
            features:torch.Tensor = self.backbone(x)
            projections:torch.Tensor = self.proj(features)
            return features.view(b,v,-1),projections.view(b,v,-1)
        else:
            features:torch.Tensor = self.backbone(x)
            projections:torch.Tensor = self.proj(features)
            return features,projections


class ViTEncoder(nn.Module):
    def __init__(self, proj_dim=128,img_size:int=224):
        super().__init__()
        self.backbone = timm.create_model(
            "vit_small_patch8_224",
            pretrained=False,
            num_classes=512,
            drop_path_rate=0.1,
            img_size=img_size,
        )
        self.proj = MLP(512, [2048, 2048, proj_dim], norm_layer=nn.BatchNorm1d)

    def forward(self, x):
        if x.ndim == 5:
            # Perform inference on views
            b,v,c,h,w = x.size()
            x = x.view(b*v,c,h,w)
            features:torch.Tensor = self.backbone(x)
            projections:torch.Tensor = self.proj(features)
            return features.view(b,v,-1),projections.view(b,v,-1)
        else:
            features:torch.Tensor = self.backbone(x)
            projections:torch.Tensor = self.proj(features)
            return features,projections
```

</details>

As you can see the only difference is that i keep the batch first.
Consequently, i did this implementation of SigReg :
<details>
<summary>SigReg Implementation</summary>
I followed the paper implementation but i get the exact same result than your implementation when i transpose the input (batch and views).
Bellow my implementations : (SIGRegGithub correspond to your minimal example, i just added the generator for reproductibility.

```python
import torch
import torch.distributed as dist

class SIGReg(torch.nn.Module):
    def __init__(self, tmax:int=3,knots:int=17,num_slices:int=256):
        super().__init__()
        self.register_buffer("exp_f", None)
        self.register_buffer("t", None)
        self.num_slices = num_slices
        t = torch.linspace(0, tmax, knots, dtype=torch.float32)
        # We divide tmax by the number of intervals.
        self.t = t
        # Theorical CF for N(0,1) and Gauss windows
        self.exp_f = torch.exp(-t.square()/2) # T

    def forward(self, x:torch.Tensor,step:int=None):
        """Compute the SIGReg

        Args:
            x (torch.Tensor): (B,V,K) tensor (B = Batch,V = Number of views, K = Dimension)

        Returns:
            Errors: (num_slices) Errors compared to normal distribution on each slice.
        """
        B = x.size(0)
        K = x.size(-1)
        # We create num_slices random directions.
        # Theres direction are unit size
        g = None
        if step is not None:
            g = torch.Generator(device=x.device)
            g.manual_seed(step)
        A = torch.randn(K, self.num_slices, device=x.device,generator=g)
        A = A.div_(A.norm(p=2, dim=0))
        # Empirical CF
        x_t = (x @ A).unsqueeze(-1)*self.t # (B,V,K,1) -> (B,V,K,T)
        # We average on batch dimension
        # In the paper : https://arxiv.org/abs/2511.08544
        # Dimensions are [V,B,K,T] and the loop is done on V
        # It means N in the paper -> Batch
        # Then, because here :
        # Dimensions are [B,V,K,T] -> we average first on B -> dim 0
        ecf = (1j*x_t).exp().mean(0) # -> (V,K,T)
        if dist.is_initialized():
            dist.all_reduce(ecf,op=dist.ReduceOp.AVG)
            B=B*dist.get_world_size()
        # Weighted L2 Norm
        err = (ecf-self.exp_f).abs().square().mul(self.exp_f) # (V,K,T)@T -> (V,K,T)
        # We integrate
        # Like before, we multiply by B (N) in the paper
        # But we multiply by 2 because of the symetry of the integral.
        T = 2*torch.trapz(err,self.t,dim=-1)*B # (V,K)
        # We average the 
        # Average on Views (V).
        return T.mean()

class SIGRegGitHub(torch.nn.Module):
    def __init__(self, knots=17):
        super().__init__()
        t = torch.linspace(0, 3, knots, dtype=torch.float32)
        dt = 3 / (knots - 1)
        weights = torch.full((knots,), 2 * dt, dtype=torch.float32)
        weights[[0, -1]] = dt
        window = torch.exp(-t.square() / 2.0)
        self.register_buffer("t", t)
        self.register_buffer("phi", window)
        self.register_buffer("weights", weights * window)

    def forward(self, proj,step=0):
        g = None
        if step is not None:
            g = torch.Generator(device=proj.device)
            g.manual_seed(step)
        A = torch.randn(proj.size(-1), 256, device=x.device,generator=g)
        A = A.div_(A.norm(p=2, dim=0))
        x_t = (proj @ A).unsqueeze(-1) * self.t
        err = (x_t.cos().mean(-3) - self.phi).square() + x_t.sin().mean(-3).square()
        statistic = (err @ self.weights) * proj.size(-2)
        return statistic.mean()

```
As you can see, it's completely consistant :
<img width="605" height="204" alt="Image" src="https://github.com/user-attachments/assets/7cc5a401-9f60-4696-915e-f282491eaec0" />
</details>

The training loop is consistent as well with what you proposed in your minimal example. I did a lot of tries with different hyper parameters but the accuracy is stucked near 0.5.

<img width="330" height="308" alt="Image" src="https://github.com/user-attachments/assets/bca0fad0-0de1-486b-9178-425fc15c3a06" />

During training, i see that the gradient in the model is not null, but the sigreg is stucked arround 18.

<img width="1021" height="640" alt="Image" src="https://github.com/user-attachments/assets/43a82c61-0381-4706-ad7c-dad50cbb030b" />

The configuration looks like that : 

```python
@dataclass
class TrainingConfig:
    # Dataset
    train_dataset: str = field(default="Multimodal-Fatima/Imagenette_train", metadata={"description": "Name of the dataset to use", "example": "timm/mini-imagenet"})
    val_dataset: str = field(default="Multimodal-Fatima/Imagenette_validation", metadata={"description": "Name of the dataset to use", "example": "timm/mini-imagenet"})
    num_classes:int = field(default=10)
    global_size:int = field(default=128)
    # Outputs
    output_dir: str = field(default="./outputs/", metadata={"description": "Directory to save outputs", "example": "./outputs/"})
    # Training
    model_name:str = field(default='resnet18')
    num_globals:int = field(default=8)
    num_locals:int = field(default=0)
    num_training_steps:int = field(default=100_000)
    val_frequency:int = field(default=500)
    batch_size: int = field(default=256, metadata={"description": "Batch size for training", "example": 32})
    learning_rate: float = field(default=5e-3, metadata={"description": "Learning rate for the optimizer", "example": 5e-4})
    probe_learning_rate: float = field(default=1e-3, metadata={"description": "Learning rate for the probe optimizer", "example": 1e-3})
    warmup_steps: int = field(default=500, metadata={"description": "Number of warmup steps for learning rate scheduler", "example": 1})
    gradient_accumulation_steps: int = field(default=1, metadata={"description": "Number of steps to accumulate gradients", "example": 1})
    weight_decay: float = field(default=5e-4, metadata={"description": "Weight decay for the optimizer", "example": 5e-4})
    suffix:str = field(default="ImageNette")
    # Infra
    mixed_precision:str = field(default="bf16",metadata={"description": "Mixed precision to use while training", "example":"no/fp16/bf16"})
    # LeJEPA
    projection_dim: int = field(default=512, metadata={"description": "Dimension of the projection head output", "example": 256})
    lambda_lejepa:float = field(default=0.02, metadata={"description":"Ratio Loss vs Regularisation"})
```

I did some experimentation arround Sigreg in a notebook with random gaussian values and i saw that convergence to 0 is almost impossible (it stuck arround 1). I don't know if it's normal or not. But while training it's really hard to have at the same time the Sigreg converging and the inv_loss converging at the same time. (Even with playing on lambda coefficient). I got the exact same issue with resnet18 and vit8.

I tried to remove weight_decay on bias and batchnorm layers but without positive effect.
If it was stucking at zero, i would definitely think something's wrong from my side but at the beggining training seems efficient. Just it stuck after several iterations.

Do you see something obvious i could have miss ?

Thank you very much for your help :)
